File: ROSM3/ex01/full_name_package/src/service_name.cpp
#include "rclcpp/rclcpp.hpp"
#include "full_name_package/srv/summ_full_name.hpp"

using namespace std::placeholders;

class ServiceNode : public rclcpp::Node
{
public:
  ServiceNode() : Node("service_name")
  {
    service_ = this->create_service<full_name_package::srv::SummFullName>(
      "SummFullName",
      std::bind(&ServiceNode::handle_service, this, _1, _2)
    );
    // Вставить сюда:
    RCLCPP_INFO(this->get_logger(), "Service SummFullName is ready!");
  }

  void handle_service(
    const std::shared_ptr<full_name_package::srv::SummFullName::Request> request,
    std::shared_ptr<full_name_package::srv::SummFullName::Response> response)
  {
    // Вставить сюда:
    RCLCPP_INFO(this->get_logger(), "Incoming request: %s %s %s",
      request->last_name.c_str(),
      request->first_name.c_str(),
      request->middle_name.c_str()
    );
    response->full_name = request->last_name + " " +
                          request->first_name + " " +
                          request->middle_name;
  }

private:
  rclcpp::Service<full_name_package::srv::SummFullName>::SharedPtr service_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ServiceNode>());
  rclcpp::shutdown();
  return 0;
}

File: ROSM3/ex01/ex01/ros2_ws/src/service_full_name/service_full_name/service_name.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from service_full_name.srv import SummFullName

class ServiceName(Node):

    def __init__(self):
        super().__init__('service_name') # инициализирует узел с именем 'service_name'
        self.srv = self.create_service(SummFullName, 'SummFullName', self.handle_summ_full_name) #  создаёт сервис с именем 'SummFullName', типом SummFullName и callback-функцией handle_summ_full_name

    def handle_summ_full_name(self, request, response): #  функция, вызываемая при каждом запросе к сервису
        full_name = f"{request.last_name} {request.name} {request.first_name}" # формирует полное имя из полей запроса.
        self.get_logger().info(f"Received: {request.last_name}, {request.name}, {request.first_name}")
        response.full_name = full_name # заполняет поле ответа полным именем
        self.get_logger().info(f"Sending back: {full_name}")
        return response # возвращает заполненный объект ответа клиенту


def main(args=None):
    rclpy.init(args=args)
    service = ServiceName()
    rclpy.spin(service)
    rclpy.shutdown()

if __name__ == '__main__':
    main()
